File: app/src/main/java/com/mainstreetcode/teammate/fragments/main/GameFragment.java
package com.mainstreetcode.teammate.fragments.main;

import android.content.Context;
import android.os.Bundle;
import android.support.annotation.DrawableRes;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.annotation.StringRes;
import android.support.design.chip.Chip;
import android.support.design.widget.BaseTransientBottomBar;
import android.support.v7.app.AlertDialog;
import android.support.v7.util.DiffUtil;
import android.transition.AutoTransition;
import android.transition.TransitionManager;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;

import com.mainstreetcode.teammate.R;
import com.mainstreetcode.teammate.adapters.StatAdapter;
import com.mainstreetcode.teammate.adapters.UserAdapter;
import com.mainstreetcode.teammate.adapters.viewholders.ChoiceBar;
import com.mainstreetcode.teammate.adapters.viewholders.EmptyViewHolder;
import com.mainstreetcode.teammate.adapters.viewholders.GameViewHolder;
import com.mainstreetcode.teammate.baseclasses.BottomSheetController;
import com.mainstreetcode.teammate.baseclasses.MainActivityFragment;
import com.mainstreetcode.teammate.model.Competitor;
import com.mainstreetcode.teammate.model.Event;
import com.mainstreetcode.teammate.model.Game;
import com.mainstreetcode.teammate.model.Identifiable;
import com.mainstreetcode.teammate.model.Stat;
import com.mainstreetcode.teammate.model.User;
import com.mainstreetcode.teammate.util.AppBarListener;
import com.mainstreetcode.teammate.util.ErrorHandler;
import com.mainstreetcode.teammate.util.ScrollManager;
import com.mainstreetcode.teammate.util.ViewHolderUtil;
import com.mainstreetcode.teammate.viewmodel.gofers.GameGofer;
import com.tunjid.androidbootstrap.core.abstractclasses.BaseFragment;

import java.util.List;
import java.util.concurrent.atomic.AtomicBoolean;

/**
 * Lists {@link Event games}
 */

public final class GameFragment extends MainActivityFragment
        implements UserAdapter.AdapterListener {

    private static final String ARG_GAME = "game";

    private Game game;
    private List<Identifiable> items;

    private boolean hasChoiceBar;
    private GameGofer gofer;
    private AtomicBoolean editableStatus;
    private AtomicBoolean privilegeStatus;
    private GameViewHolder gameViewHolder;

    private Chip refereeChip;

    public static GameFragment newInstance(Game game) {
        GameFragment fragment = new GameFragment();
        Bundle args = new Bundle();

        args.putParcelable(ARG_GAME, game);
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    @SuppressWarnings("ConstantConditions")
    public String getStableTag() {
        String superResult = super.getStableTag();
        Game tempGame = getArguments().getParcelable(ARG_GAME);

        return (tempGame != null)
                ? superResult + "-" + tempGame.hashCode()
                : superResult;
    }

    @Override
    @SuppressWarnings("ConstantConditions")
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setHasOptionsMenu(true);

        Bundle arguments = getArguments();
        game = arguments.getParcelable(ARG_GAME);
        items = statViewModel.getModelList(game);
        editableStatus = new AtomicBoolean();
        privilegeStatus = new AtomicBoolean();
        gofer = gameViewModel.gofer(game);
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View rootView = inflater.inflate(R.layout.fragment_game, container, false);
        View appBar = rootView.findViewById(R.id.app_bar);
        refereeChip = rootView.findViewById(R.id.referee_chip);
        gameViewHolder = new GameViewHolder(appBar, ignored -> {});
        gameViewHolder.bind(game);

        scrollManager = ScrollManager.withRecyclerView(rootView.findViewById(R.id.model_list))
                .withEmptyViewholder(new EmptyViewHolder(rootView, R.drawable.ic_stat_white_24dp, R.string.no_stats))
                .withAdapter(new StatAdapter(items, stat -> showFragment(StatEditFragment.newInstance(stat))))
                .withRefreshLayout(rootView.findViewById(R.id.refresh_layout), this::refresh)
                .withEndlessScrollCallback(() -> fetchStats(false))
                .addScrollListener((dx, dy) -> updateFabForScrollState(dy))
                .withInconsistencyHandler(this::onInconsistencyDetected)
                .withLinearLayoutManager()
                .build();

        AppBarListener.with().appBarLayout(rootView.findViewById(R.id.app_bar))
                .offsetDiffListener(gameViewHolder::animate).create();

        scrollManager.setViewHolderColor(R.color.dark_grey);

        refereeChip.setCloseIconResource(R.drawable.ic_close_24dp);
        refereeChip.setOnCloseIconClickListener(v -> onRemoveRefereeClicked());
        refereeChip.setOnClickListener(v -> onRefereeClicked());
        rootView.findViewById(R.id.home_thumbnail).setOnClickListener(view -> showCompetitor(game.getHome()));
        rootView.findViewById(R.id.away_thumbnail).setOnClickListener(view -> showCompetitor(game.getAway()));
        rootView.findViewById(R.id.score).setOnClickListener(view -> showFragment(GameEditFragment.newInstance(game)));
        rootView.findViewById(R.id.date).setOnClickListener(view -> showFragment(EventEditFragment.newInstance(game)));

        return rootView;
    }

    @Override
    public void onPrepareOptionsMenu(Menu menu) {
        super.onPrepareOptionsMenu(menu);
        menu.findItem(R.id.action_delete_game).setVisible(gofer.canDelete(userViewModel.getCurrentUser()));
        menu.findItem(R.id.action_end_game).setVisible(showsFab());
        menu.findItem(R.id.action_event).setVisible(true);
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        inflater.inflate(R.menu.fragment_game, menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.action_end_game:
                endGameRequest();
                return true;
            case R.id.action_event:
                showFragment(EventEditFragment.newInstance(game));
                return true;
            case R.id.action_delete_game:
                Context context = getContext();
                if (context == null) return true;
                new AlertDialog.Builder(context).setTitle(getString(R.string.game_delete_prompt))
                        .setPositiveButton(R.string.yes, (dialog, which) -> deleteGame())
                        .setNegativeButton(R.string.no, (dialog, which) -> dialog.dismiss())
                        .show();
                return true;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public void onResume() {
        super.onResume();
        fetchGame();
        fetchStats(true);
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        refereeChip = null;
    }

    @Override
    public void togglePersistentUi() {
        updateFabIcon();
        setFabClickListener(this);
        setToolbarTitle(getString(R.string.game_stats));
        requireActivity().invalidateOptionsMenu();
        super.togglePersistentUi();
    }

    @Override
    @StringRes
    protected int getFabStringResource() { return R.string.stat_add; }

    @Override
    @DrawableRes
    protected int getFabIconResource() { return R.drawable.ic_add_white_24dp; }

    @Override
    public boolean showsFab() {
        return editableStatus.get() && !game.isEnded();
    }

    @Override
    public void onClick(View v) {
        if (v.getId() == R.id.fab) showFragment(StatEditFragment.newInstance(Stat.empty(game)));
    }

    @Override
    public void onUserClicked(User item) {
        hideKeyboard();
        refereeChip.postDelayed(() -> {
            hideBottomSheet();
            refereeChip.postDelayed(() -> addRefereeRequest(item), 150);
        }, 200);
    }

    private void updateGame() {
        disposables.add(gofer.save().subscribe(ignored -> onGameUpdated(), defaultErrorHandler));
    }

    private void fetchGame() {
        disposables.add(gofer.fetch().subscribe(ignored -> onGameUpdated(), defaultErrorHandler));
    }

    private void refresh() {
        fetchGame();
        disposables.add(statViewModel.refresh(game).subscribe(GameFragment.this::onStatsFetched, defaultErrorHandler));
    }

    private void updateStatuses() {
        disposables.add(statViewModel.isPrivilegedInGame(game)
                .subscribe(privilegeStatus::set, defaultErrorHandler));
        disposables.add(statViewModel.canEditGameStats(game).doOnSuccess(editableStatus::set)
                .subscribe(ignored -> togglePersistentUi(), defaultErrorHandler));

        if (game.competitorsDeclined())
            scrollManager.updateForEmptyList(R.drawable.ic_stat_white_24dp, R.string.no_competitor_declined);

        else if (game.competitorsNotAccepted())
            scrollManager.updateForEmptyList(R.drawable.ic_stat_white_24dp, R.string.no_competitor_acceptance);

        else scrollManager.updateForEmptyList(R.drawable.ic_stat_white_24dp, R.string.no_stats);
    }

    private void onRemoveRefereeClicked() {
        if (!privilegeStatus.get()) return;
        game.getReferee().update(User.empty());
        toggleProgress(true);
        updateGame();
    }

    private void onRefereeClicked() {
        User referee = game.getReferee();
        boolean hasReferee = !referee.isEmpty();
        if (hasReferee) showFragment(UserEditFragment.newInstance(referee));
        else if (privilegeStatus.get()) {
            BaseFragment fragment = UserSearchFragment.newInstance();
            fragment.setTargetFragment(this, R.id.request_user_pick);
            showBottomSheet(BottomSheetController.Args.builder()
                    .setMenuRes(R.menu.empty)
                    .setFragment(fragment)
                    .build());
        }
    }

    private void endGameRequest() {
        new AlertDialog.Builder(requireActivity())
                .setTitle(R.string.game_end_request)
                .setMessage(R.string.game_end_prompt)
                .setPositiveButton(R.string.yes, (dialog, which) -> {
                    game.setEnded(true);
                    toggleProgress(true);
                    disposables.add(gofer.save().subscribe(ignored -> onGameUpdated(), defaultErrorHandler));
                })
                .setNegativeButton(R.string.cancel, (dialog, which) -> dialog.dismiss())
                .show();
    }

    private void addRefereeRequest(User user) {
        new AlertDialog.Builder(requireActivity())
                .setTitle(R.string.game_add_referee_title)
                .setMessage(R.string.game_add_referee_prompt)
                .setPositiveButton(R.string.yes, (dialog, which) -> {
                    game.getReferee().update(user);
                    toggleProgress(true);
                    updateGame();
                })
                .setNegativeButton(R.string.cancel, (dialog, which) -> dialog.dismiss())
                .show();
    }

    private void fetchStats(boolean fetchLatest) {
        if (fetchLatest) scrollManager.setRefreshing();
        else toggleProgress(true);

        disposables.add(statViewModel.getMany(game, fetchLatest).subscribe(this::onStatsFetched, defaultErrorHandler));
        updateStatuses();
    }

    private void onStatsFetched(DiffUtil.DiffResult result) {
        toggleProgress(false);
        scrollManager.onDiff(result);
    }

    private void onGameUpdated() {
        disposables.add(gofer.watchForChange().subscribe(changed -> requireActivity().invalidateOptionsMenu(), ErrorHandler.EMPTY));
        gameViewHolder.bind(game);
        toggleProgress(false);
        checkCompetitor();
        updateStatuses();
        bindReferee();
    }

    private void respond(boolean accept, Competitor competitor) {
        toggleProgress(true);
        disposables.add(competitorViewModel.respond(competitor, accept)
                .subscribe(ignored -> {
                    if (accept) fetchGame();
                    else toggleProgress(false);
                }, defaultErrorHandler));
    }

    private void deleteGame() {
        disposables.add(gofer.remove().subscribe(this::onGameDeleted, defaultErrorHandler));
    }

    private void onGameDeleted() {
        showSnackbar(getString(R.string.game_deleted));
        removeEnterExitTransitions();
        requireActivity().onBackPressed();
    }

    private void checkCompetitor() {
        if (hasChoiceBar) return;
        if (!roleViewModel.privilegedInGame(game)) return;

        Competitor competitor;
        if (game.getHome().hasNotResponded()) competitor = game.getHome();
        else if (game.getAway().hasNotResponded()) competitor = game.getAway();
        else competitor = null;

        if (competitor == null || competitor.isEmpty() || competitor.isAccepted()) return;

        User currentUser = userViewModel.getCurrentUser();
        if (game.betweenUsers() && !currentUser.equals(competitor.getEntity())) return;

        hasChoiceBar = true;
        showChoices(choiceBar -> choiceBar.setText(getString(R.string.game_accept))
                .setPositiveText(getText(R.string.accept))
                .setNegativeText(getText(R.string.decline))
                .setPositiveClickListener(v -> respond(true, competitor))
                .setNegativeClickListener(v -> respond(false, competitor))
                .addCallback(new BaseTransientBottomBar.BaseCallback<ChoiceBar>() {
                    public void onDismissed(ChoiceBar shown, int event) { hasChoiceBar = false; }
                }));
    }

    private void bindReferee() {
        ViewGroup root = (ViewGroup) getView();
        if (root == null) return;

        int size = refereeChip.getResources().getDimensionPixelSize(R.dimen.double_margin);
        User referee = game.getReferee();
        TransitionManager.beginDelayedTransition(root, new AutoTransition().addTarget(refereeChip));

        refereeChip.setCloseIconVisible(!referee.isEmpty() && privilegeStatus.get());
        refereeChip.setText(referee.isEmpty()
                ? getString(privilegeStatus.get() && !game.isEnded()
                ? R.string.game_choose_referee : R.string.game_no_referee)
                : getString(R.string.game_referee, referee.getName()));

        disposables.add(ViewHolderUtil.fetchRoundedDrawable(refereeChip.getContext(), referee.getImageUrl(), size)
                .subscribe(refereeChip::setChipIcon, ErrorHandler.EMPTY));
    }
}

File: app/src/main/java/com/mainstreetcode/teammate/fragments/main/GameFragment.kt
/*
 * MIT License
 *
 * Copyright (c) 2019 Adetunji Dahunsi
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package com.mainstreetcode.teammate.fragments.main

import android.os.Bundle
import android.transition.AutoTransition
import android.transition.TransitionManager
import android.view.Menu
import android.view.MenuItem
import android.view.View
import androidx.appcompat.app.AlertDialog
import androidx.core.os.bundleOf
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.snackbar.BaseTransientBottomBar
import com.mainstreetcode.teammate.R
import com.mainstreetcode.teammate.adapters.Shell
import com.mainstreetcode.teammate.adapters.statAdapter
import com.mainstreetcode.teammate.adapters.viewholders.ChoiceBar
import com.mainstreetcode.teammate.adapters.viewholders.EmptyViewHolder
import com.mainstreetcode.teammate.adapters.viewholders.GameViewHolder
import com.mainstreetcode.teammate.baseclasses.TeammatesBaseFragment
import com.mainstreetcode.teammate.baseclasses.removeSharedElementTransitions
import com.mainstreetcode.teammate.databinding.FragmentGameBinding
import com.mainstreetcode.teammate.model.Competitor
import com.mainstreetcode.teammate.model.Event
import com.mainstreetcode.teammate.model.Game
import com.mainstreetcode.teammate.model.ListState
import com.mainstreetcode.teammate.model.Stat
import com.mainstreetcode.teammate.model.User
import com.mainstreetcode.teammate.util.AppBarListener
import com.mainstreetcode.teammate.util.ErrorHandler
import com.mainstreetcode.teammate.util.ScrollManager
import com.mainstreetcode.teammate.util.fetchRoundedDrawable
import com.mainstreetcode.teammate.viewmodel.gofers.GameGofer
import com.tunjid.androidx.recyclerview.diff.Differentiable
import java.util.concurrent.atomic.AtomicBoolean

/**
 * Lists [games][Event]
 */

class GameFragment : TeammatesBaseFragment(R.layout.fragment_game), Shell.UserAdapterListener {

    private lateinit var game: Game
    private lateinit var items: List<Differentiable>

    private var hasChoiceBar: Boolean = false
    private val editableStatus: AtomicBoolean = AtomicBoolean()
    private val privilegeStatus: AtomicBoolean = AtomicBoolean()
    private var gameViewHolder: GameViewHolder? = null
    private var binding: FragmentGameBinding? = null

    private lateinit var gofer: GameGofer

    override val showsFab: Boolean get() = !bottomSheetDriver.isBottomSheetShowing && editableStatus.get() && !game.isEnded

    override val stableTag: String
        get() {
            val superResult = super.stableTag
            val tempGame = arguments!!.getParcelable<Game>(ARG_GAME)

            return if (tempGame != null) superResult + "-" + tempGame.hashCode()
            else superResult
        }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        val arguments = arguments
        game = arguments!!.getParcelable(ARG_GAME)!!
        items = statViewModel.getModelList(game)
        gofer = gameViewModel.gofer(game)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) = FragmentGameBinding.bind(view).run {
        super.onViewCreated(view, savedInstanceState)
        defaultUi(
                toolbarTitle = getString(R.string.game_stats),
                toolBarMenu = R.menu.fragment_game,
                fabText = R.string.stat_add,
                fabIcon = R.drawable.ic_add_white_24dp
        )

        gameViewHolder = GameViewHolder(appBar) {}
        gameViewHolder?.bind(game)

        scrollManager = ScrollManager.with<RecyclerView.ViewHolder>(view.findViewById(R.id.model_list))
                .withPlaceholder(EmptyViewHolder(view, R.drawable.ic_stat_white_24dp, R.string.no_stats))
                .withAdapter(statAdapter(::items) { navigator.push(StatEditFragment.newInstance(it)) })
                .withRefreshLayout(view.findViewById(R.id.refresh_layout)) { this@GameFragment.refresh() }
                .withEndlessScroll { fetchStats(false) }
                .addScrollListener { _, dy -> updateFabForScrollState(dy) }
                .withInconsistencyHandler(this@GameFragment::onInconsistencyDetected)
                .withLinearLayoutManager()
                .build()

        scrollManager.setViewHolderColor(R.attr.alt_empty_view_holder_tint)

        refereeChip.setCloseIconResource(R.drawable.ic_close_24dp)
        refereeChip.setOnCloseIconClickListener { onRemoveRefereeClicked() }
        refereeChip.setOnClickListener { onRefereeClicked() }

        homeThumbnail.setOnClickListener { showCompetitor(game.home) }
        awayThumbnail.setOnClickListener { showCompetitor(game.away) }
        score.setOnClickListener { navigator.push(GameEditFragment.newInstance(game)) }
        date.setOnClickListener { navigator.push(EventEditFragment.newInstance(game)) }
        appBar.apply { AppBarListener(this) { gameViewHolder?.animate(it) } }

        binding = this
    }

    override fun onPrepareOptionsMenu(menu: Menu) {
        super.onPrepareOptionsMenu(menu)
        menu.findItem(R.id.action_delete_game)?.isVisible = gofer.canDelete(userViewModel.currentUser)
        menu.findItem(R.id.action_end_game)?.isVisible = showsFab
        menu.findItem(R.id.action_event)?.isVisible = true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean = when (item.itemId) {
        R.id.action_end_game -> endGameRequest().let { true }
        R.id.action_event -> navigator.push(EventEditFragment.newInstance(game))
        R.id.action_delete_game -> AlertDialog.Builder(requireContext())
                .setTitle(getString(R.string.game_delete_prompt))
                .setPositiveButton(R.string.yes) { _, _ -> deleteGame() }
                .setNegativeButton(R.string.no) { dialog, _ -> dialog.dismiss() }
                .show().let { true }
        else -> super.onOptionsItemSelected(item)
    }

    override fun onResume() {
        super.onResume()
        fetchGame()
        fetchStats(true)
    }

    override fun onDestroyView() {
        super.onDestroyView()
        gameViewHolder = null
        binding = null
    }

    override fun onKeyBoardChanged(appeared: Boolean) {
        super.onKeyBoardChanged(appeared)
        if (!appeared && bottomSheetDriver.isBottomSheetShowing) bottomSheetDriver.hideBottomSheet()
    }

    override fun onClick(view: View) {
        if (view.id == R.id.fab) navigator.push(StatEditFragment.newInstance(Stat.empty(game)))
    }

    override fun onUserClicked(item: User) {
        hideKeyboard()
        binding?.refereeChip?.postDelayed({
            bottomSheetDriver.hideBottomSheet()
            binding?.refereeChip?.postDelayed({ addRefereeRequest(item) }, 150)
        }, 200)
    }

    private fun updateGame() {
        disposables.add(gofer.save().subscribe({ onGameUpdated() }, defaultErrorHandler::invoke))
    }

    private fun fetchGame() {
        disposables.add(gofer.fetch().subscribe({ onGameUpdated() }, defaultErrorHandler::invoke))
    }

    private fun refresh() {
        fetchGame()
        disposables.add(statViewModel.refresh(game).subscribe(this::onStatsFetched, defaultErrorHandler::invoke))
    }

    private fun updateStatuses() {
        disposables.add(statViewModel.isPrivilegedInGame(game)
                .subscribe(privilegeStatus::set, defaultErrorHandler::invoke))

        disposables.add(statViewModel.canEditGameStats(game).doOnSuccess(editableStatus::set)
                .subscribe({ updateUi(fabShows = showsFab) }, defaultErrorHandler::invoke))

        when {
            game.competitorsDeclined() -> scrollManager.updateForEmptyList(ListState(R.drawable.ic_stat_white_24dp, R.string.no_competitor_declined))
            game.competitorsNotAccepted() -> scrollManager.updateForEmptyList(ListState(R.drawable.ic_stat_white_24dp, R.string.no_competitor_acceptance))
            else -> scrollManager.updateForEmptyList(ListState(R.drawable.ic_stat_white_24dp, R.string.no_stats))
        }
    }

    private fun onRemoveRefereeClicked() {
        if (!privilegeStatus.get()) return
        game.referee.update(User.empty())
        transientBarDriver.toggleProgress(true)
        updateGame()
    }

    private fun onRefereeClicked() {
        val referee = game.referee
        val hasReferee = !referee.isEmpty

        if (hasReferee) navigator.push(UserEditFragment.newInstance(referee))
        else if (privilegeStatus.get()) bottomSheetDriver.showBottomSheet(
                requestCode = R.id.request_user_pick,
                fragment = UserSearchFragment.newInstance()
        )
    }

    private fun endGameRequest() {
        AlertDialog.Builder(requireActivity())
                .setTitle(R.string.game_end_request)
                .setMessage(R.string.game_end_prompt)
                .setPositiveButton(R.string.yes) { _, _ ->
                    game.isEnded = true
                    transientBarDriver.toggleProgress(true)
                    disposables.add(gofer.save().subscribe({ onGameUpdated() }, defaultErrorHandler::invoke))
                }
                .setNegativeButton(R.string.cancel) { dialog, _ -> dialog.dismiss() }
                .show()
    }

    private fun addRefereeRequest(user: User) {
        AlertDialog.Builder(requireActivity())
                .setTitle(R.string.game_add_referee_title)
                .setMessage(R.string.game_add_referee_prompt)
                .setPositiveButton(R.string.yes) { _, _ ->
                    game.referee.update(user)
                    transientBarDriver.toggleProgress(true)
                    updateGame()
                }
                .setNegativeButton(R.string.cancel) { dialog, _ -> dialog.dismiss() }
                .show()
    }

    private fun fetchStats(fetchLatest: Boolean) {
        if (fetchLatest) scrollManager.setRefreshing()
        else transientBarDriver.toggleProgress(true)

        disposables.add(statViewModel.getMany(game, fetchLatest).subscribe(this::onStatsFetched, defaultErrorHandler::invoke))
        updateStatuses()
    }

    private fun onStatsFetched(result: DiffUtil.DiffResult) {
        transientBarDriver.toggleProgress(false)
        scrollManager.onDiff(result)
    }

    private fun onGameUpdated() {
        disposables.add(gofer.watchForChange().subscribe({ updateUi(toolbarInvalidated = true) }, ErrorHandler.EMPTY::invoke))
        gameViewHolder?.bind(game)
        transientBarDriver.toggleProgress(false)
        checkCompetitor()
        updateStatuses()
        bindReferee()
    }

    private fun respond(accept: Boolean, competitor: Competitor) {
        transientBarDriver.toggleProgress(true)
        disposables.add(competitorViewModel.respond(competitor, accept)
                .subscribe({
                    if (accept) fetchGame()
                    else transientBarDriver.toggleProgress(false)
                }, defaultErrorHandler::invoke))
    }

    private fun deleteGame() {
        disposables.add(gofer.remove().subscribe(this::onGameDeleted, defaultErrorHandler::invoke))
    }

    private fun onGameDeleted() {
        transientBarDriver.showSnackBar(getString(R.string.game_deleted))
        removeSharedElementTransitions()
        requireActivity().onBackPressed()
    }

    private fun checkCompetitor() {
        if (hasChoiceBar) return
        disposables.add(roleViewModel.hasPendingCompetitor(game).subscribe({ competitor ->
            hasChoiceBar = true
            transientBarDriver.showChoices { choiceBar ->
                choiceBar.setText(getString(R.string.game_accept))
                        .setPositiveText(getText(R.string.accept))
                        .setNegativeText(getText(R.string.decline))
                        .setPositiveClickListener(View.OnClickListener { respond(true, competitor) })
                        .setNegativeClickListener(View.OnClickListener { respond(false, competitor) })
                        .addCallback(object : BaseTransientBottomBar.BaseCallback<ChoiceBar>() {
                            override fun onDismissed(shown: ChoiceBar?, event: Int) {
                                hasChoiceBar = false
                            }
                        })
            }
        }, ErrorHandler.EMPTY::invoke))
    }

    private fun bindReferee() = binding?.apply {
        val size = resources.getDimensionPixelSize(R.dimen.double_margin)
        val referee = game.referee
        TransitionManager.beginDelayedTransition(innerCoordinator, AutoTransition().addTarget(refereeChip))

        refereeChip.isCloseIconVisible = !referee.isEmpty && privilegeStatus.get()
        refereeChip.text = when {
            referee.isEmpty -> getString(when {
                privilegeStatus.get() && !game.isEnded -> R.string.game_choose_referee
                else -> R.string.game_no_referee
            })
            else -> getString(R.string.game_referee, referee.name)
        }

        disposables.add(fetchRoundedDrawable(requireContext(), referee.imageUrl, size)
                .subscribe({ refereeChip?.chipIcon = it }, ErrorHandler.EMPTY::invoke))
    }

    companion object {

        private const val ARG_GAME = "game"

        fun newInstance(game: Game): GameFragment = GameFragment().apply { arguments = bundleOf(ARG_GAME to game) }
    }
}
